piptools/utils.py

```diff
@@ -522,3 +531,126 @@ def parse_requirements_from_wheel_metadata(
             markers=parts.markers,
             extras=parts.extras,
         )
+
+
+def callback_config_file_defaults(
+    ctx: click.Context, param: click.Parameter, value: str | None
+) -> str | None:
+    """
+    Returns the path to the config file with defaults being used, or `None` if no such file is
+    found.
+
+    Defaults for `click.Command` parameters should be override-able in a config file. `pip-tools`
+    will use the first file found, searching in this order: an explicitly given config file, a
+    `.pip-tools.toml`, a `pyproject.toml` file. Those files are searched for in the same directory
+    as the requirements input file.
+    """
+    if value is None:
+        config_file = select_config_file(ctx.params.get("src_files", ()))
+        if config_file is None:
+            return None
+    else:
+        config_file = Path(value)
+
+    try:
+        config = parse_config_file(config_file)
+    except OSError as e:
+        raise click.FileError(
+            filename=str(config_file), hint=f"Could not read '{config_file}': {e}"
+        )
+    except ValueError as e:
+        raise click.FileError(
+            filename=str(config_file), hint=f"Could not parse '{config_file}': {e}"
+        )
+
+    if not config:
+        return None
+
+    defaults: dict[str, Any] = ctx.default_map.copy() if ctx.default_map else {}
+    defaults.update(config)
+
+    ctx.default_map = defaults
```

Comment:
Upon looking into this more, I'm now convinced that this does not belong here at all. It should happen externally, rather than being tightly coupled to the path argument validation and normalization. Being a side effect only makes the control flow more entangled.

Reply:
I think this bears more discussion.

The order of precedence in determining the value for a particular option is: 1. An explicit declaration at runtime via a flag; 2. A default specified in a config file; 3. A default hard-coded in pip-tools. The configuration file's values need to override the built-in defaults but not any explicitly declared values. By the time the `cli()` function is called, explicit declarations and hard-coded defaults have already been resolved with one another, and it becomes convoluted to try to determine for which options the config file's values should be operative. The same holds true for decorators.

By using an `is_eager=True` option with a callback, we have the opportunity to selectively override hard-coded defaults with new config-file-specified defaults before any commandline explicit options are handled. I don't see a simple place to inject this order of precedence otherwise, and from Click documentation, that follows the intent of the `is_eager` flag:

> Sometimes, you want a parameter to completely change the execution flow. [ ... ] In such cases, you need two concepts: eager parameters and a callback.

If you can imagine an approach using a decorator or in the `cli()` function that can nimbly achieve this order of precedence, I'm open to hearing it.

Reply:
Okay, now I understand the challenges better. Marking this thread as resolved.